ufl/corealg/dag_traverser.py

```diff
@@ -0,0 +1,123 @@
+"""Base class for dag traversers."""
+
+from abc import ABC, abstractmethod
+from functools import singledispatchmethod, wraps
+
+from ufl.classes import Expr
+
+
+class DAGTraverser(ABC):
+    """Base class for dag traversers.
+
+    Args:
+        compress: If True, ``result_cache`` will be used.
+        visited_cache: cache of intermediate results; expr -> r = self.process(expr, ...).
+        result_cache: cache of result objects for memory reuse, r -> r.
+
+    """
+
+    def __init__(self, compress=True, visited_cache=None, result_cache=None) -> None:
```

Comment:
```suggestion
    def __init__(self, compress: bool=True, visited_cache=None, result_cache=None) -> None:
```
Currently I do not think there is any test in `ufl` that has a `*args` as a non-empty argument. i.e. is always equal to `args=()`. Could you add a test that has `args` being non-empty?

Reply:
Should the type-hint for `visited_cache` be `dict[tuple[Expr, ...], Expr]|None`, and similarly for `result_cache`?

Reply:
Added typehint.

That is right. It seems that we only need `args = ()` in `apply_derivatives.py`. We will need non-empty `args` for instance when we want to pass down the state in `apply_restrictions.py`; underneath a `Restricted` node `restricted='+'` or `restricted='-'`, otherwise `restricted=None`, where the method signature would look like:
```
@process.register(...)
def _(self, o, restricted):
    ...
```
We could remove `*args` for this PR.